themes/src/main/resources/theme/base/login/webauthn-register.ftl

```diff
@@ -0,0 +1,136 @@
+
+        let attestationConveyancePreference = "${attestationConveyancePreference}";
+        if(attestationConveyancePreference !== 'not specified') publicKey.attestation = attestationConveyancePreference;
+
+        let authenticatorSelection = {};
+        let isAuthenticatorSelectionSpecified = false;
+
+        let authenticatorAttachment = "${authenticatorAttachment}";
+        if(authenticatorAttachment !== 'not specified') {
+            authenticatorSelection.authenticatorAttachment = authenticatorAttachment;
+            isAuthenticatorSelectionSpecified = true;
+        }
+
+        let requireResidentKey = "${requireResidentKey}";
+        if(requireResidentKey !== 'not specified') {
+            if(requireResidentKey === 'Yes')
+                authenticatorSelection.requireResidentKey = true;
+            else
+                authenticatorSelection.requireResidentKey = false;
+            isAuthenticatorSelectionSpecified = true;
+        }
+
+        let userVerificationRequirement = "${userVerificationRequirement}";
+        if(userVerificationRequirement !== 'not specified') {
+            authenticatorSelection.userVerification = userVerificationRequirement;
+            isAuthenticatorSelectionSpecified = true;
+        }
+
+        if(isAuthenticatorSelectionSpecified) publicKey.authenticatorSelection = authenticatorSelection;
+
+        let createTimeout = ${createTimeout};
+        if(createTimeout != 0) publicKey.timeout = createTimeout * 1000;
+
+        let excludeCredentialIds = "${excludeCredentialIds}";
+        let excludeCredentials = getExcludeCredentials(excludeCredentialIds);
+        if (excludeCredentials.length > 0) publicKey.excludeCredentials = excludeCredentials;
+
+        navigator.credentials.create({publicKey})
+            .then(function(result) {
+                window.result = result;
+                let clientDataJSON = result.response.clientDataJSON;
+                let attestationObject = result.response.attestationObject;
+                let publicKeyCredentialId = result.rawId;
+
+                $("#clientDataJSON").val(base64url.encode(new Uint8Array(clientDataJSON), { pad: false }));
+                $("#attestationObject").val(base64url.encode(new Uint8Array(attestationObject), { pad: false }));
+                $("#publicKeyCredentialId").val(base64url.encode(new Uint8Array(publicKeyCredentialId), { pad: false }));
+
+                let initLabel = "WebAuthn Authenticator (Default Label)";
+                let labelResult = window.prompt("Please input your registered authenticator's label", initLabel);
+                if (labelResult === null) labelResult = initLabel;
+                $("#authenticatorLabel").val(labelResult);
+
+                $("#register").submit();
+
+            })
+            .catch(function(err) {
+                $("#error").val(err);
+                $("#register").submit();
+
+            });
+
+        function getPubKeyCredParams(signatureAlgorithms) {
```

Comment:
As for using single bean class, I would like to try it as you said.

As for moving the business logic inside UI template onto java class, I'm afraid it takes a relatively lot of time compared with other tasks.

I've already confirmed it before as follows :

- create the business logic implemented endpoint in the REST manner that receives the user's input about the options for WebAuthn API,  and returns the option itself for WebAuthn API (navigator.credentials.create(), .get())

- create the UI template which receives the user's input and access the above endpoint  in the asynchronous manner, and call WebAuthn API (navigator.credentials.create(), .get()).

However, I'm not sure how to secure this endpoint like other endpoints (e.g. login-actions). Therefore I've decided that this feature has not been included in this PR.

Reply:
Yep, ok. That's fine.

I personally wouldn't try to use REST API for this. At least not in this PR. Binding the parameters to the freemarker is fine IMO.

Reply:
Yes, at this PR, I'll try to bind parameters to the freemarker as you said.

Reply:
I guess one option could be to bind freemarker objects as strings? For example in java code, you can create the "representation" object and then convert it to JSON String with the KEycloak JsonSerialization.writeValueAsString and bind this string to the freemarker (or addd it as a property to freemarker bean). Then in freemarker template JS code, you can call "JSON.parse" to get back the JSON object.

Just a minor thing, not a blocker for this PR IMO.

Reply:
I've try it but had some problem. Some special characters in the string processed by JsonSerialization.writeValueAsString has been replaced with HTML character references (e.g. 
 " -> &quot) automatically. Therefore, JSON.parse failed. It seems that keycloak automatically sanitizes any variable binded into Freemarker template.

Maybe the following seems work but it need some conversion work :

[Java]
- The representation of PublicKeyCredentialCreationOptions -> JSON as String
- JSON as String -> Byte Array
- Byte Array -> Base64URL Encoded String

[Javascript]
- Base64URL Encoded Sting -> Byte Array
- Byte Array -> JSON as String
- JSON as String -> Object in Javascript

Reply:
I see. Then maybe we rather keep as it is rather than add more layers of conversion (array of bytes, Base64URL etc), which can also affect performance a bit etc?

Reply:
Yes, I would like to work on the other task (especially, not using UserModel's attribute for public key credential's metadata) and will back to this issue later on. What do you think about it?

Reply:
+1